unittests/savanna_disaster_recovery.cpp

```diff
@@ -0,0 +1,270 @@
+   C.remove_state();
+   C.overwrite_fsi(fsi);
+   C.open_from_snapshot(snapshot);
+   A.push_blocks_to(C);
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2); // all 4 finalizers should be back voting
+   BOOST_REQUIRE(!C.is_head_missing_finalizer_votes());                        // let's make sure of that
+} FC_LOG_AND_RETHROW()
+
+// ---------------------------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE(recover_killed_node_with_deleted_fsi, savanna_cluster::cluster_t) try {
+   auto& A=_nodes[0]; auto& C=_nodes[2];
+
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2);
+   auto snapshot = C.snapshot();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2);
+   C.close();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2); // lib still advances with 3 finalizers
+   C.remove_state();
+   C.remove_fsi();
+   C.open_from_snapshot(snapshot);
+   A.push_blocks_to(C);
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2); // all 4 finalizers should be back voting
+   BOOST_REQUIRE(!C.is_head_missing_finalizer_votes());                        // let's make sure of that
+} FC_LOG_AND_RETHROW()
+
+// ---------------------------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE(recover_killed_node_while_retaining_fsi, savanna_cluster::cluster_t) try {
+   auto& A=_nodes[0]; auto& C=_nodes[2];
+
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2);
+   auto snapshot = C.snapshot();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2);
+   C.close();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2); // lib still advances with 3 finalizers
+   C.remove_state();
+   C.open_from_snapshot(snapshot);
+   A.push_blocks_to(C);
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(2);  }), 2); // all 4 finalizers should be back voting
+   BOOST_REQUIRE(!C.is_head_missing_finalizer_votes());                        // let's make sure of that
+} FC_LOG_AND_RETHROW()
+
+
+// ---------------------------------------------------------------------------------------------------
+//                               All but one finalizers go down
+// ---------------------------------------------------------------------------------------------------
+
+// ---------------------------------------------------------------------------------------------------
+BOOST_FIXTURE_TEST_CASE(nodes_go_down, savanna_cluster::cluster_t) try {
+   auto& A=_nodes[0]; auto& B=_nodes[1]; auto& C=_nodes[2]; auto& D=_nodes[3];
+   std::array<savanna_cluster::node_t*, 3> failing_nodes { &B, &C, &D };
+
+   for (auto& N : failing_nodes) N->close();
+   BOOST_REQUIRE_EQUAL(A.lib_advances_by([&]() { A.produce_blocks(4);  }), 1); // lib stalls with 3 finalizers down
```

Comment:
Why 1 not 0?

Reply:
Because there was one QC in flight (that the nodes had already voted on) to be included into the next block.

Reply:
Maybe add a comment there.

Reply:
done